File: examples/vewprocessing/example1/convert_linestrings_to_vewstrings.py
# %%
import os
import yaml
import numpy as np
import geopandas as gpd
from shapely.geometry import Point, LineString
import matplotlib.pyplot as plt
from scipy.spatial import cKDTree
from adcircpy import AdcircMesh

# %%
# Paths to the input/output files
f14file = "/home/sbunya/GitHub/adcircutils/adcircutils/vewchannel/examples/example1/fort.14"
vewlocation_geofile = "/home/sbunya/GitHub/adcircutils/adcircutils/vewchannel/examples/example1/example1_vew_locations.shp"
vewfile = "/home/sbunya/GitHub/adcircutils/adcircutils/vewchannel/examples/example1/vewstings.yaml"

# Parameters
dist_max = 10.0  # Maximum distance for nearest neighbor search in meters

# %%
# Read the shapefile
vewlocation_gdf = gpd.read_file(vewlocation_geofile)
print(vewlocation_gdf)

# %%
# Load the mesh file
mesh = AdcircMesh.open(f14file)

# %%
# Create nodestrings along the lines in the geospatial file
x = mesh.nodes.x.values
y = mesh.nodes.y.values
tri = mesh.triangles - 1

# Create a KDTree for fast nearest neighbor search
tree = cKDTree(np.c_[x, y])

neighs = mesh.node_neighbors
neighs = [list(neighs[i]) for i in range(len(neighs))]

nodestrings = []

for iline, line in enumerate(vewlocation_gdf.geometry):
    slx, sly = line.coords[0][0], line.coords[0][1]
    elx, ely = line.coords[-1][0], line.coords[-1][1]
    
    # Query the KDTree for the nearest neighbor
    distance, nearest_node_index = tree.query([slx, sly])

    # Get the coordinates of the nearest node
    nearest_node_x = x[nearest_node_index]
    nearest_node_y = y[nearest_node_index]

    ni = nearest_node_index
    xi = nearest_node_x
    yi = nearest_node_y
    distance_to_endpoint = 1e10
    
    nodestring = [nearest_node_index]
    
    ipos = 0
    while ipos < len(line.coords) - 1:
        # Get the points in the line segment
        point0 = line.coords[ipos]
        point1 = line.coords[ipos + 1]
        xl0, yl0 = point0[0], point0[1]
        xl1, yl1 = point1[0], point1[1]

        line_segment = LineString([(xl0, yl0), (xl1, yl1)])
        
        neigh = neighs[ni]
        neigh = [n for n in neigh if n not in nodestring]

        xnei = [x[n] for n in neigh]
        ynei = [y[n] for n in neigh]
        
        # Compute the distances from the line to the points defined by xnei and ynei
        distances = [line_segment.distance(Point(xn, yn)) for xn, yn in zip(xnei, ynei)]
        min_distance = np.min(distances)
        
        if min_distance > dist_max:
            ipos += 1
            continue
        
        min_distance_index = np.argmin(distances)
        ni = neigh[min_distance_index]
        xi = x[ni]
        yi = y[ni]
        
        nodestring.append(int(ni))
        
        distance_to_endpoint = np.sqrt((elx - xi)**2 + (ely - yi)**2)
        
        if distance_to_endpoint < dist_max:
            break

    xi, yi = x[nodestring[0]], y[nodestring[0]]
    distance_to_endpoint = np.sqrt((elx - xi)**2 + (ely - yi)**2)
    if distance_to_endpoint < dist_max:
        nodestring.append(int(nodestring[0]))
        
    nodestrings.append(nodestring)

# %%
# Visualize the nodestrings
fig, ax = plt.subplots()
for nodestring in nodestrings:
    x_nodes = [x[n] for n in nodestring]
    y_nodes = [y[n] for n in nodestring]
    ax.plot(x_nodes, y_nodes, marker='o')
    print(x_nodes)
ax.set_xlabel('X Coordinate')
ax.set_ylabel('Y Coordinate')

# %%
# Assign values to each node in the nodestrings
nodestrings_with_values = []
for nodestring in nodestrings:
    nodestring_with_value = \
        [{'node': node+1, 
          'x': float(mesh.nodes.x.loc[node]), 'y': float(mesh.nodes.y.loc[node]),
          'bank_elevation': 1.5 if mesh.nodes.y.loc[node] > 500.0 else -4.0,
          'bank_mannings_n': 0.03,
          } for node in nodestring]
    nodestrings_with_values.append(nodestring_with_value)

# Prepare the data for YAML output
data = {'vewstrings': nodestrings_with_values}

# Print the YAML output
print(yaml.dump(data, sort_keys=False))

# Write the YAML output to a file
with open(vewfile, 'w') as f:
    yaml.dump(data, f, sort_keys=False)

File: examples/vewprocessing/example1/convert_linestrings_to_vewstrings.m
% Paths to the input/output files
basedir = 'H:\mnt\sshfs\hatteras\home\sbunya\GitHub\adcircutils\adcircutils\vewchannel\examples\example1';
f14file = fullfile(basedir, 'fort.14');
vewlocation_geofile = fullfile(basedir, 'example1_vew_locations.shp');
vewfile = fullfile(basedir, 'vewstings.yaml');

dist_max = 10.0; % Maximum distance for nearest neighbor search in meters

% Read the shapefile
vewlocation_gdf = shaperead(vewlocation_geofile);
disp(vewlocation_gdf);

% Load the mesh file
mesh = msh(f14file); % Custom function to load ADCIRC mesh

x = mesh.nodes(:,1);
y = mesh.nodes(:,2);
tri = mesh.;

% Create KDTree for nearest neighbor search
tree = KDTreeSearcher([x, y]);

nodestrings = {};

for i = 1:length(vewlocation_gdf)
    line = vewlocation_gdf(i).X;
    slx = line(1);
    sly = line(2);
    elx = line(end-1);
    ely = line(end);
    
    [nearest_node_index, distance] = knnsearch(tree, [slx, sly]);
    
    if distance > dist_max
        continue;
    end
    
    nodestring = nearest_node_index;
    
    ipos = 1;
    while ipos < length(line) - 1
        xl0 = line(ipos);
        yl0 = line(ipos+1);
        xl1 = line(ipos+2);
        yl1 = line(ipos+3);
        
        % Find neighboring nodes
        neighbors = find_neighbors(mesh, nearest_node_index); % Custom function
        
        xnei = x(neighbors);
        ynei = y(neighbors);
        
        distances = arrayfun(@(i) distance_point_line([xnei(i), ynei(i)], [xl0, yl0], [xl1, yl1]), 1:length(neighbors));
        [min_distance, min_idx] = min(distances);
        
        if min_distance > dist_max
            ipos = ipos + 2;
            continue;
        end
        
        nearest_node_index = neighbors(min_idx);
        nodestring = [nodestring, nearest_node_index];
        
        if norm([elx, ely] - [x(nearest_node_index), y(nearest_node_index)]) < dist_max
            break;
        end
    end
    
    if norm([elx, ely] - [x(nodestring(1)), y(nodestring(1))]) < dist_max
        nodestring = [nodestring, nodestring(1)];
    end
    
    nodestrings{end+1} = nodestring;
end

% Visualization
figure; hold on;
for i = 1:length(nodestrings)
    plot(x(nodestrings{i}), y(nodestrings{i}), '-o');
end
xlabel('X Coordinate'); ylabel('Y Coordinate');

% Assign values to each node
nodestrings_with_values = cell(size(nodestrings));
for i = 1:length(nodestrings)
    nodestrings_with_values{i} = arrayfun(@(node) struct('node', node, 'x', x(node), 'y', y(node), ...
        'bank_elevation', 1.5 * (y(node) > 500) + -4.0 * (y(node) <= 500), 'bank_mannings_n', 0.03), ...
        nodestrings{i}, 'UniformOutput', false);
end

data = struct('vewstrings', {nodestrings_with_values});
yaml_write(data, vewfile); % Custom function to write YAML